model_compression_toolkit/qunatizers_infrastructure/pytorch/quantize_wrapper.py

```diff
@@ -0,0 +1,123 @@
+
+if FOUND_TORCH:
+    import torch
+    import torch.nn as nn
+
+    DISPATCHER = "dispatcher"
+    LAYER = "layer"
+    STEPS = "optimizer_step"
+    TRAINING = "training"
+
+
+    def _weight_name(name: str) -> str:
+        """Extracts the weight name from the full TensorFlow variable name.
+
+        For example, returns 'kernel' for 'dense_2/kernel:0'.
+
+        Args:
+          name: TensorFlow variable name.
+
+        Returns:
+          Extracted weight name.
+        """
+        return name.split(':')[0].split('/')[-1]
+
+
+    class PytorchQuantizationWrapper(nn.Module):
+        def __init__(self,
+                     module: nn.Module,
+                     dispatcher: NodeQuantizationDispatcher):
+            """
+            Pytorch Quantization Wrapper takes a pytorch module and dispatcher and infer a quantized module.
+
+            Args:
+                layer: A pytorch module.
+                dispatcher: A node quantization dispatcher.
+            """
+            super().__init__()
+            self.dispatcher = dispatcher
+            self.add_module('layer', module)
+            self._weight_vars = []
+            for name, quantizer in self.dispatcher.weight_quantizers.items():
+                weight = getattr(self.layer, name)
+                quantizer.initialize_quantization(weight.shape,
+                                                  name+'_quant', self)
+
+                self._weight_vars.append((name, weight, quantizer))
+
+
+        def set_quantize_weights(self, quantized_weights: dict):
+            """
+            This function update layer weights after quantization.
+
+            Args:
+                quantized_weights: a dict of weight to update
+
+            Returns: None
+
+            """
+            for weight_attr in self.dispatcher.weight_quantizers.keys():
+                weight = quantized_weights.get(weight_attr)
+                current_weight = getattr(self.layer, weight_attr)
+                if current_weight.shape != weight.shape:
+                    Logger.error(
+                        f"Existing layer weight shape {current_weight.shape} is incompatible with provided weight shape {weight.shape}")  # pragma: no cover
+
+                setattr(self.layer, weight_attr, torch.nn.Parameter(weight))
+
+        def forward(self, x: torch.Tensor) -> torch.Tensor:
+            """
+            PytorchQuantizationWrapper forward functions
+            Args:
+                inputs: Input tensors to specified layer
+                training: a boolean stating if layer is in training mode.
+                **kwargs:
+
+            Returns: tensors that simulate quantized layer.
+
+            """
+
+            # Quantize all weights, and replace them in the underlying layer.
+
+            quantized_weights = {}
+            for name, unquantized_weight, quantizer in self._weight_vars:
+                quantized_weight = quantizer.forward(unquantized_weight, self.training)
```

Comment:
I'm not sure, but is calling "forward" is the correct way to call for a Pytorch Module inference? shouldn't it be implicit, i.e., just `quantizer(unquantized_weight, self.training)`?

Reply:
you are right. apparently, "quantizer" is not a nn.Module... so the "forward" method is just a name. I'll change it to "__call__ " method and then we can drop the "forward" which is just confusing.